### PR TITLE
support high-order derivatives for some operators.

### PR DESCRIPTION
**Pull request by yoonlee888**
- [x] tanh
- [ ] sin
- [ ] cos

### CODE REVIEW

File: oneflow/core/autograd/gradient_funcs/split_like.cpp [hunk 1]
```diff
@@ -59,7 +59,7 @@ Maybe<void> SplitLike::Capture(SplitLikeCaptureState* ctx, const TensorTuple& in
 
 Maybe<void> SplitLike::Apply(const SplitLikeCaptureState* ctx, const TensorTuple& out_grads,
                              TensorTuple* in_grads) const {
-  in_grads->resize(1);
+  in_grads->resize(out_grads.size() + 1);
```
Comment (wyg1997) [inline oneflow/core/autograd/gradient_funcs/split_like.cpp:62]: split_like inputs 的个数等于 outputs 的个数+1

File: oneflow/core/autograd/gradient_funcs/tanh.cpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  }
+
+  Maybe<void> Apply(const TanhCaptureState* ctx, const TensorTuple& out_grads,
+                    TensorTuple* in_grads) const override {
+    in_grads->resize(1);
+    if (!ctx->x_requires_grad) { return Maybe<void>::Ok(); }
+    const auto& y = ctx->SavedTensors().at(0);
+    const auto& a = functional::Mul(y, y);
+    const auto& aa = functional::ScalarSub(1, JUST(a));
+    in_grads->at(0) = JUST(functional::Mul(out_grads.at(0), JUST(aa)));
```
Comment (wyg1997) [inline oneflow/core/autograd/gradient_funcs/tanh.cpp:46]: 这里可以用 squeese_function 优化一下

File: oneflow/core/functional/impl/array_functor.cpp [hunk 3]
```diff
@@ -2346,7 +2346,7 @@ class SplitLikeFunctor {
     JUST(attrs.SetAttr<int64_t>("axis", axis));
     TensorTuple inputs(like.size() + 1);
     inputs[0] = x;
-    for (int i = 0; i < like.size(); ++i) { inputs[i + 1] = like[i]; }
+    for (int i = 0; i < like.size(); ++i) { inputs[i + 1] = JUST(like[i]->detach()); }
```
Comment (wyg1997) [inline oneflow/core/functional/impl/array_functor.cpp:2349]: detach 要放在 functor 里

File: oneflow/core/functional/impl/higher_derivative_functor.cpp [hunk 4]
```diff
@@ -57,11 +57,24 @@ class CosGradGradFunctor {
   }
 };
 
+class TanhGradGradFunctor {
+ public:
+  Maybe<Tensor> operator()(const std::shared_ptr<Tensor>& x,
+                           const std::shared_ptr<Tensor>& dydx) const {
```
Comment (hengzi) [inline oneflow/core/functional/impl/higher_derivative_functor.cpp:63]: 可以改成 sequence_function 方式吗
